core-bundle/src/Security/Voter/DataContainer/ContentCompositionVoter.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\Voter\DataContainer;
+
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Routing\Page\PageRegistry;
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\CoreBundle\Security\DataContainer\CreateAction;
+use Contao\CoreBundle\Security\DataContainer\UpdateAction;
+use Contao\LayoutModel;
+use Contao\PageModel;
+use Contao\StringUtil;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\CacheableVoterInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\VoterInterface;
+
+/**
+ * @internal
+ */
+class ContentCompositionVoter implements VoterInterface, CacheableVoterInterface
+{
+    public function __construct(
+        private readonly ContaoFramework $framework,
+        private readonly PageRegistry $pageRegistry,
+    ) {
+    }
+
+    public function supportsAttribute(string $attribute): bool
+    {
+        return ContaoCorePermissions::DC_PREFIX.'tl_article' === $attribute;
+    }
+
+    public function supportsType(string $subjectType): bool
+    {
+        return \in_array($subjectType, [CreateAction::class, UpdateAction::class], true);
+    }
+
+    public function vote(TokenInterface $token, $subject, array $attributes): int
+    {
+        foreach ($attributes as $attribute) {
+            if (
+                (!$subject instanceof CreateAction && !$subject instanceof UpdateAction)
```

Comment:
This should also be checked outside of the `foreach` loop.

Reply:
technically yes, but all voter handle each attribute separately. There is always only one attribute, it's just an array for legacy reasons (also see https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php#L54-L57).